lib/storage/metadata/mongoclient/MongoClientInterface.js

```diff
@@ -965,6 +965,85 @@ class MongoClientInterface {
         ], cb);
     }
 
+    /**
+     * gets object metadata for a list of objects
+     * @param {String} bucketName bucket name
+     * @param {Array} objects array of objects
+     * @param {Object} log logger
+     * @param {Function} callback callback
+     * @return {undefined}
+     */
+    getObjects(bucketName, objects, log, callback) {
+        const c = this.getCollection(bucketName);
+        let vFormat = null;
+        if (!Array.isArray(objects)) {
+            return callback(errors.InternalError.customizeDescription('objects must be an array'));
+        }
+        // Function to process each document
+        const processDoc = (doc, objName, params, key, cb) => {
+            const versionIdValue = params && params.versionId ? params.versionId : undefined;
+            if (!doc && versionIdValue) {
+                // If no document and a version ID is provided, return an error.
+                return cb(null, {
+                    err: errors.NoSuchKey,
+                    doc: null,
+                    versionId: versionIdValue,
+                    key,
+                });
+            }
+            // If no master found then object is either non existent or last
+            // version is delete marker
+            if (!doc || doc.value.isPHD) {
+                return this.getLatestVersion(c, objName, vFormat, log, (err, _doc) => cb(null, {
+                    err,
+                    doc: _doc || null,
+                    versionId: versionIdValue,
+                    key,
+                }));
+            }
+            MongoUtils.unserialize(doc.value);
+            return cb(null, {
+                err: null,
+                doc: doc.value,
+                versionId: versionIdValue,
+                key,
+            });
+        };
+        return this.getBucketVFormat(bucketName, log, (err, _vFormat) => {
+            if (err) {
+                return callback(err);
+            }
+            vFormat = _vFormat;
+            // Create keys, maintaining the context with each key
+            const keysAndObjects = objects.map(({ key: objName, params }) => {
+                const _key = params && params.versionId
+                    ? formatVersionKey(objName, params.versionId, vFormat)
+                    : formatMasterKey(objName, vFormat);
+                return { key: _key, objName, params };
+            });
+            // Extract keys and find documents
+            const keys = keysAndObjects.map(o => o.key);
```

Comment:
is this "map" really useful?
it seems we really just need the list of keys (for mongo), and later we can iterate over `objects` directly...

seems lots of data structure are created in here (`keysAndObjects`, `keys`, `docByKey`) which may end up costing memory/gc : and maybe 'cost' more than re-formatting the key twice...

not sure if this is an issue or if we need to improve this

Reply:
So far no issues with the memory with testing... We already greatly improve it thanks to the batching, instead of doing 1000s of calls to MongoDB...

Reply:
Here is my tentative: https://github.com/scality/Arsenal/pull/2134/commits/cc75b89115a7a8d4f26417fde5103f80421bedef#diff-9c207befbd8f4d79a3b6e3d7d143444e804ca1684236b8fd9fa0f9e03c53477aR976
It removes one of the maps